### DISCUSSION TITLE
Should we keep the proxy for login?

### DISCUSSION THREAD

**Discussion by alandtse** [alandtse/tesla#23] [General]
We can either:
1. Keep the oauth proxy which requires a workaround and may break easily in the future.
2. Switch to only accepting a refresh token that has be generated by a third-party app. However, we are reliant on those being able to defeat Tesla's login security.

I do not intend to do both since it's the worst of both worlds.

Thoughts?

**Comment by jherby2k** [comment]
It's really up to you and what you're willing to maintain. Going with 2 could make it into the core integration right?

**Reply by alandtse** [reply]
No. Both are rejected by core and we're taking it out from core since there's no path forward without either.

**Reply by jherby2k** [reply]
Crummy. I'd say go with 2 and save yourself a recurring headache. Better to focus on features than login BS

**Comment by hbar98** [comment]
If you are going to do 1, then I'll work on getting logs. If you are going to do 2, then I won't worry about it right now.

I'm an utter neophyte at this level, so the only suggestion that I'd make is to go with whichever option that you can correct the fastest or easiest should something go wrong.

**Comment by omelhus** [comment]
I think the easiest way is to just use the refresh token and the external apps. The proxy will be a never ending game of cat and mouse.

**Comment by sillyfrog** [comment]
Re 2: I'm not sure how an external app works, but is it possible to get the tokens from an existing browser session? (eg: copy the relevant cookies?) Or even better, most browsers have copy as `cURL` or `HAR`, get these details, and paste them into an input box, and use python to parse out the relevant tokens (if it could work).

**Reply by omelhus** [reply]
You just use Auth App for Tesla on iOS or Tesla Tokens on Android to generate a token. It's really easy and the input in the config flow will just be the refresh token.

For the time being the sso tokens on the account page at tesla.com is using an old endpoint (v1) to generate the tokens and cannot be used to generate new v3 tokens.

**Reply by jameshiscott** [reply]
indeed, hence its super easy from an instructions on what to do for new uesrs

**Comment by ybizeul** [comment]
I'm down for the external app tokens, seems to be a reasonable alternative, considering how everyone seems to be committed in making it complicated.

**Comment by jameshiscott** [comment]
I would agree with option 2. The support level for new users is much lower than having to change host files etc

**Comment by nicholashead** [comment]
I believe tokens would be the best long-term thing. It may be worth working with https://github.com/adriankumpf/teslamate or other open source projects to see on the most "correct" way?

**Reply by alandtse** [reply]
[Teslamate](https://github.com/adriankumpf/teslamate/issues/1945#issuecomment-913684723) is using tokens now.

**Comment by omelhus** [comment]
Another benefit of using the token apps is that the issue with tokens is no longer a concern for the component, and many already rely on the token apps today.

The token app devs can focus on the auth and we can get heated steering wheels before winter👍

**Reply by hbar98** [reply]
I just want to be able to view my solar gateway in HA.

**Reply by alandtse** [reply]
> Another benefit of using the token apps is that the issue with tokens is no longer a concern for the component, and many already rely on the token apps today.
> 
> The token app devs can focus on the auth and we can get heated steering wheels before winter👍

Yah let's go with tokens. Do you have time to get your PR into this repo? You'll just need to throw out the existing config flow and replace it. I won't have time until the weekend. The only caution may be the variables we save into the config entry are slightly differently named.

**Reply by omelhus** [reply]
I’ll see what I can do in the morning. Should be an easy fix.

**Comment by blair287** [comment]
I'm confused if other apps like tesla for wear os etc just need an email and password why is there such an issue for HA? When they are still all working OK.

**Reply by alandtse** [reply]
They're obviously smarter than those of us working in Python. 

But assuming you're asking a serious question, all of the apps are trying to imitate Tesla's official apps because Tesla can't block their own apps. This includes requiring the app respond in certain ways including with technology designed to stop third party access. At a certain point, Tesla can tighten the controls so fake apps can't access the servers. From Tesla's perspective, we're the same as malware. The better question is why Tesla has taken this long to stop us.

**Reply by nicholashead** [reply]
@alandtse just to throw this out there - I recently had issues with my Model 3 charge port not closing - and the tech straight up refused to work on my car unless I changed my password first, to break all "unauthorized third party apps" from using their API. Because his theory was my charge port was broke because of Teslamate/etc. and he was seeing "strange readings". Total BS - but that was their stance (and in the end the charge port issue had nothing to do with API). 

It's really unfortunate that Tesla isn't doing a proper API approach here. I'd expect it to work like every other OAuth/etc type system - where they list out apps that have third party access and you can revoke at any time. But alas.

**Reply by omelhus** [reply]
When you’ve authenticated with Tesla they redirect you to something like auth.tesla.com/void/callback with a code as a query parameter. 

If we could have a listener for current url in home assistant on mobile we could do the same thing and hijack the callback and use the code to get the tokens.

**Reply by alandtse** [reply]
That's what the oauth proxy used to do, but I'm not interested in writing a mobile app to do that if someone else already did. We're either in the cat and mouse game or we're not.

There was a [programmatic solution](https://github.com/timdorr/tesla-api/discussions/390#discussioncomment-1304576) but it requires subscribing to the 2captcha service where you pay per catpcha solved. I am not interested in doing that because I find captcha solving services distasteful since I believe they're primary business is for cracking people's accounts in an automated fashion.

**Reply by blair287** [reply]
> They're obviously smarter than those of us working in Python.
> 
> But assuming you're asking a serious question, all of the apps are trying to imitate Tesla's official apps because Tesla can't block their own apps. This includes requiring the app respond in certain ways including with technology designed to stop third party access. At a certain point, Tesla can tighten the controls so fake apps can't access the servers. From Tesla's perspective, we're the same as malware. The better question is why Tesla has taken this long to stop us.

It was a genuine query I was just wondering why it was such an issue with HA I didn't realise tesla doesn't like 3rd party apps I though they had a api for that reason.

Wasn't trying to be funny.

One of the reasons I bought a tesla was because I could automate things via HA.

**Reply by omelhus** [reply]
I’ve submitted a PR to @alandtse repo, so you’ll be back at automating your Tesla in no time 👍

**Reply by blair287** [reply]
> I’ve submitted a PR to @alandtse repo, so you’ll be back at automating your Tesla in no time 👍

Will it require the integration to be reset up every month when the token expires?

Awesome work it's working now :)

**Reply by omelhus** [reply]
> Will it require the integration to be reset up every month when the token expires?

Shouldn’t be necessary. It’s the same mechanism that keeps you logged into the app.
